### PR TITLE
CT: Add BlobHashes to TransactionContext

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds a slice of hashes to `BlockContext`, and the corresponding generator. This is to store a list of hashes, which can later on be queried by `BLOBHASH` instruction. But considering the implementation of that instruction requires a new Condition and maybe more, I think it might be best to merge this changes first and deal with the Condition and Rules in separate PRs.

This PR contributes to #468 

EDIT: noticed that `blobbasefee` was not added to `adapter_test` so it is added in this PR as well even if not strictly related.

### CODE REVIEW

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/utils/adapter_test.go [hunk 1]
```diff
@@ -313,9 +313,9 @@ func TestAdapter_ParameterConversion(t *testing.T) {
 
 	for name, property := range properties {
 		t.Run(name, func(t *testing.T) {
-			state := st.State{}
-			property.set(&state)
-			params := ToVmParameters(&state)
+			state := st.NewState(st.NewCode([]byte{}))
+			property.set(state)
+			params := ToVmParameters(state)
```
Comment (HerbertJordan) [inline go/ct/utils/adapter_test.go:318]: This change looks unrelated ... why is it needed?

File: go/ct/st/block_context.go [hunk 2]
```diff
@@ -32,6 +34,44 @@ type BlockContext struct {
...
+	if len(b.BlobHashes) != len(other.BlobHashes) {
+		isEqual = false
+	} else {
+		for i, hash := range b.BlobHashes {
+			if !slices.Equal(hash[:], other.BlobHashes[i][:]) {
+				isEqual = false
+				break
+			}
+		}
+	}
```
Comment (HerbertJordan) [inline go/ct/st/block_context.go:70]: This should be the same as
```suggestion
	isEqual = isEqual && (slices.Equal(b.BlobHashes, other.BlobHashes))
```

File: go/ct/st/block_context.go [hunk 3]
```diff
@@ -45,6 +85,16 @@ func (b *BlockContext) Diff(other *BlockContext) []string {
 		ret = append(ret, blockDifference+fmt.Sprintf("blob base fee: %v vs %v\n", b.BlobBaseFee, other.BlobBaseFee))
 	}
 
+	if len(b.BlobHashes) != len(other.BlobHashes) {
+		ret = append(ret, blockDifference+fmt.Sprintf("blob hashes: %v vs %v\n", b.BlobHashes, other.BlobHashes))
+	} else {
+		for i, hash := range b.BlobHashes {
+			if !slices.Equal(hash[:], other.BlobHashes[i][:]) {
+				ret = append(ret, blockDifference+fmt.Sprintf("blob hash: %v vs %v\n", hash, other.BlobHashes[i]))
```
Comment (HerbertJordan) [inline go/ct/st/block_context.go:93]: Please include `i` in the output.

File: go/ct/st/block_context.go [hunk 4]
```diff
@@ -45,6 +85,16 @@ func (b *BlockContext) Diff(other *BlockContext) []string {
 		ret = append(ret, blockDifference+fmt.Sprintf("blob base fee: %v vs %v\n", b.BlobBaseFee, other.BlobBaseFee))
 	}
 
+	if len(b.BlobHashes) != len(other.BlobHashes) {
+		ret = append(ret, blockDifference+fmt.Sprintf("blob hashes: %v vs %v\n", b.BlobHashes, other.BlobHashes))
+	} else {
+		for i, hash := range b.BlobHashes {
+			if !slices.Equal(hash[:], other.BlobHashes[i][:]) {
```
Comment (HerbertJordan) [inline go/ct/st/block_context.go:92]: Simpler:
```suggestion
			if hash != other.BlobHashes[i] {
```

File: go/ct/st/block_context.go [hunk 5]
```diff
@@ -32,6 +34,44 @@ type BlockContext struct {
 	TimeStamp   uint64     // Block's timestamp in unix time in seconds
 }
 
+// NewBlockContext creates a new block context
+func NewBlockContext() *BlockContext {
+	return &BlockContext{BlobHashes: []vm.Hash{}}
```
Comment (HerbertJordan) [inline go/ct/st/block_context.go:39]: The handling of the `BlockContext` could be simplified if `nil` is considered a valid value for the list.

File: go/ct/st/block_context_test.go [hunk 6]
```diff
@@ -74,3 +76,53 @@ func TestBlockContext_String(t *testing.T) {
...
+	b2 := b1.Clone()
+	if !b1.Equal(b2) {
+		t.Error("Cloned block context is not equal")
+	}
+}
+
+func TestBlockContext_ClonesAreIndependent(t *testing.T) {
+	b1 := BlockContext{}
+	b2 := b1.Clone()
+	b2.BlockNumber = 1
```
Comment (HerbertJordan) [inline go/ct/st/block_context_test.go:91]: This does not cover the one field that is not trivially copyable.

File: go/ct/st/block_context.go [hunk 7]
```diff
@@ -23,6 +24,7 @@ import (
 type BlockContext struct {
 	BaseFee     U256       // Base fee in wei
 	BlobBaseFee U256       // Blob Base fee in wei
+	BlobHashes  []vm.Hash  // List of blob hashes
```
Comment (HerbertJordan) [inline go/ct/st/block_context.go:27]: I have not looked into this in detail, but as I understand the Blob system right now, blobs should be associated to the `CallContext`, not the `BlockContext`. Each transaction in a block has its own set of blobs.

File: go/ct/gen/block_context.go [hunk 8]
```diff
@@ -175,32 +176,38 @@ func (b *BlockContextGenerator) bindVariablesInRangeConstraints(resultingBlockNu
...
 
 	chainId := RandU256(rnd)
 	coinbase, err := RandAddress(rnd)
 	if err != nil {
-		return st.BlockContext{}, err
+		return &st.BlockContext{}, err
+	}
+
+	blobHashes := []vm.Hash{}
+	blobHashesCount := rnd.Intn(st.MaxDataSize)
```
Comment (HerbertJordan) [inline go/ct/gen/block_context.go:208]: That is unreasonably large (MaxDataSize = 1024). Ethereum, right now, has a limit of up to 6 Blobs per transaction. So this limit should be single-digit.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: go/ct/st/block_context.go (hunk 7 above)
Comment (HerbertJordan) [reply go/ct/st/block_context.go:27]: Thinking about it, it is neither part of the `BlockContext` nor of the `CallContext`. It is part of the `TransactionContext`, which is a new thing that could be added.

This would also move the generation challenge of the the constraint solver out of the `BlockContext` generator, avoiding adding additional complexity to it. Wdyt?

Comment (facuMH) [reply go/ct/st/block_context.go:27]: well, since the the origin of `BlockContext` and `CallContext` it was mentioned both of them had info that was actually `TransactionContext`. Should we define more fields to be moved from either of those to a `TransactionContext` ? 
That should probably a separate PR. Maybe a follow-up PR and this one just adds `BlobHashes`, which are later on moved to the `TransactionContext` ?

File: go/ct/utils/adapter_test.go (hunk 1 above)
Comment (facuMH) [reply go/ct/utils/adapter_test.go:318]: because just calling `st.State{}` does not allocate for underlying structures, such as `Storage`, `Accounts` or `BlockContext`, I noticed `set` functions do this, so I'll revert this change and add the necessary calls to `NewBlockContext`

**Review by HerbertJordan — COMMENTED** [review]
Please re-add me as a reviewer when the conflicts are resolved, the CI builds have passed, and the PR is converted from a draft to a pull request.

File: go/ct/st/state_test.go [hunk 9]
```diff
@@ -47,7 +47,7 @@ func getNewFilledState() *State {
 	s.Logs.AddLog([]byte{4, 5, 6}, NewU256(21), NewU256(22))
 	s.CallContext = CallContext{AccountAddress: vm.Address{0x01}}
 	s.BlockContext = BlockContext{BlockNumber: 1}
-	s.TransactionContext = TransactionContext{OriginAddress: vm.Address{0x01}}
+	s.TransactionContext = &TransactionContext{BlobHashes: []vm.Hash{{4, 3, 2, 1}}}
```
Comment (HerbertJordan) [inline go/ct/st/state_test.go:50]: Should this add `BlobHashes` besides the `OriginAddress` or is the replacement intentional?

File: go/ct/st/transaction_context.go [hunk 10]
```diff
@@ -12,30 +12,73 @@ package st
...
+		ret = append(ret, transactionContextDiff+fmt.Sprintf("origin address: %v vs. %v\n", t.OriginAddress,
 			other.OriginAddress))
 	}
 
+	if len(t.BlobHashes) != len(other.BlobHashes) {
+		ret = append(ret, transactionContextDiff+fmt.Sprintf("blob hashes: %v vs %v\n", t.BlobHashes, other.BlobHashes))
+	} else {
+		for i, hash := range t.BlobHashes {
+			if hash != other.BlobHashes[i] {
+				ret = append(ret, transactionContextDiff+fmt.Sprintf("blob hash at location%v: %v vs %v\n", i, hash, other.BlobHashes[i]))
```
Comment (HerbertJordan) [inline go/ct/st/transaction_context.go:61]:
```suggestion
				ret = append(ret, transactionContextDiff+fmt.Sprintf("blob hash at location %v: %v vs %v\n", i, hash, other.BlobHashes[i]))
```

File: go/ct/st/transaction_context_test.go [hunk 11]
```diff
@@ -43,16 +46,68 @@ func TestTransactionContext_String(t *testing.T) {
...
+		OriginAddress: vm.Address{1, 2, 3, 4},
+		BlobHashes:    []vm.Hash{{1, 2, 3, 4}},
+	}
+
+	t2 := t1.Clone()
+
+	if !t1.Equal(t2) {
+		t.Errorf("Cloned transaction context is not equal to original")
+	}
+}
```
Comment (HerbertJordan) [inline go/ct/st/transaction_context_test.go:80]: It should test that the list of blob hashes is indeed deep-copied.

File: go/ct/st/transaction_context_test.go [hunk 12]
```diff
@@ -43,16 +46,68 @@ func TestTransactionContext_String(t *testing.T) {
...
+
+func TestTransactionContext_Equal(t *testing.T) {
+
+	tests := map[string]struct {
+		change func(*TransactionContext, *TransactionContext)
```
Comment (HerbertJordan) [inline go/ct/st/transaction_context_test.go:85]: Why the change function instead of two fields `a` and `b`?

**Review by HerbertJordan — DISMISSED** [review]

File: go/ct/st/state_test.go (hunk 9 above)
Comment (facuMH) [reply go/ct/st/state_test.go:50]: is intentional, since the idea is to change something in the structure, and in this case, changing `BlobHashes` is more interesting than an address.

File: go/ct/st/transaction_context_test.go (hunk 12 above)
Comment (facuMH) [reply go/ct/st/transaction_context_test.go:85]: I thought change function conveyed the right meaning in this case, to modify depending on the test case.

File: go/ct/st/transaction_context_test.go [hunk 13]
```diff
@@ -43,16 +47,78 @@ func TestTransactionContext_String(t *testing.T) {
...
+			t2.BlobHashes = []vm.Hash{{4}}
+		}, false},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+
+			t1 := TransactionContext{}
+			t2 := TransactionContext{}
+			test.change(&t1, &t2)
```
Comment (simonlechner) [inline go/ct/st/transaction_context_test.go:118]: You could add an equals check before applying the change, this would help with debugging if the Equals method is broken.

**Review by LuisPH3 — APPROVED** [review]

File: go/ct/st/transaction_context_test.go (hunk 12 above)
Comment (HerbertJordan) [reply go/ct/st/transaction_context_test.go:96]: A test for an `Equal` function should be about pairs of values as they are at the time of the comparison, not about the historic evolution how they become to be what hey are today -- what is implied with the name `change` and the way changes are used in other tests.

Also, using a pairs of values would make the code much shorter and simpler:
```
	tests := map[string]struct {
		a, b *TransactionContext
	}{
		"zero-values": {
			a: &TransactionContext{},
			b: &TransactionContext{},
		},
		"match": {
			a: &TransactionContext{BlobHashes: []vm.Hash{{1}}},
			b: &TransactionContext{BlobHashes: []vm.Hash{{1}}},
		},
		"mismatch": {
			a: &TransactionContext{BlobHashes: []vm.Hash{{1}}},
			b: &TransactionContext{BlobHashes: []vm.Hash{{4}}},
		},
		"mismatch-different-blob-hash-length": {
			a: &TransactionContext{BlobHashes: []vm.Hash{{1}, {2}}},
			b: &TransactionContext{BlobHashes: []vm.Hash{{4}}},
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			a, b := test.a, test.b
			if got, want := a.Eq(b), reflect.DeepEqual(a, b); got != want {
				t.Errorf("failed to compare contexts %v and %v, wanted %t, got %t", a, b, want, got)
			}
		})
	}
```

File: go/ct/st/transaction_context_test.go [hunk 14]
```diff
@@ -43,16 +47,81 @@ func TestTransactionContext_String(t *testing.T) {
...
+		t.Run(name, func(t *testing.T) {
+
+			t1 := TransactionContext{}
+			t2 := TransactionContext{}
+			if !t1.Eq(&t2) {
+				t.Error("transaction context equals is broken")
+			}
+			test.change(&t1, &t2)
+			if t1.Eq(&t2) != test.wanted {
+				t.Error("Block contexts are equal")
```
Comment (HerbertJordan) [inline go/ct/st/transaction_context_test.go:123]: This error message is wrong.

File: go/ct/st/transaction_context_test.go [hunk 15]
```diff
@@ -43,16 +47,81 @@ func TestTransactionContext_String(t *testing.T) {
...
+			t2.BlobHashes = []vm.Hash{{1}}
+		}, true},
+		"mismatch": {func(t1, t2 *TransactionContext) {
+			t1.BlobHashes = []vm.Hash{{1}}
+			t2.BlobHashes = []vm.Hash{{4}}
+		}, false},
+		"mismatch-different-blobhash-length": {func(t1, t2 *TransactionContext) {
+			t1.BlobHashes = []vm.Hash{{1}, {2}}
+			t2.BlobHashes = []vm.Hash{{4}}
+		}, false},
```
Comment (HerbertJordan) [inline go/ct/st/transaction_context_test.go:110]: These tests do not cover the `OriginAddress`.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

**Review by HerbertJordan — APPROVED** [review]

**Review by LuisPH3 — APPROVED** [review]

**Review by simonlechner — APPROVED** [review]
